### PR TITLE
STYLE: enable pylint warnings for `cell-var-from-loop`

### PR DESCRIPTION
**Pull request by roadswitcher**
- [x] addresses pylint issues with `cell-var-from-loop` as part of general pylint efforts in  #48855 

- Found one instance of the error that was an easy lift of function outside a loop
- Refactored two other locations for error based on tips from @mroeschke 

TIL `functools.partial` is a thing.

### CODE REVIEW

File: pandas/core/strings/object_array.py [hunk 1]
```diff
@@ -382,6 +382,7 @@ def _str_get_dummies(self, sep: str = "|"):
 
         for i, t in enumerate(tags2):
             pat = sep + t + sep
+            # pylint: disable-next=cell-var-from-loop
             dummies[:, i] = lib.map_infer(arr.to_numpy(), lambda x: pat in x)
         return dummies, tags2
```
Comment (MarcoGorelli) [inline pandas/core/strings/object_array.py:387]: is it possible to not turn off this check? e.g. before the for-loop, you could have some function like
```
        def _isin(test_elements, element):
            return element in test_elements
```
and then, inside the loop:
```
            dummies[:, i] = lib.map_infer(arr.to_numpy(), functools.partial(_isin, element=pat))
```

File: pandas/io/pytables.py [hunk 2]
```diff
@@ -4124,7 +4124,7 @@ def process_filter(field, filt):
                             # the levels
                             if self.is_multi_index:
                                 filt = filt.union(Index(self.levels))
-
+                            # pylint: disable-next=cell-var-from-loop
```
Comment (MarcoGorelli) [inline pandas/io/pytables.py:4127]: I think this is a good one to fix, as the code is really nested here...can you factor out `process_filter` and define it on L4108 (before the if-statement), and give it an extra argument `op`, which you can then pass in on L4128

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks for working on this, just left a couple of comments

File: pandas/core/strings/object_array.py (hunk 1 above)
Comment (roadswitcher) [reply pandas/core/strings/object_array.py:387]: Pylint agrees with you.

File: pandas/io/pytables.py (hunk 2 above)
Comment (roadswitcher) [reply pandas/io/pytables.py:4127]: Made change, waiting to see what `pre-commit run --all-files` tells me before going for `git push`.

File: pandas/io/pytables.py [hunk 3]
```diff
@@ -4106,44 +4106,42 @@ def process_axes(self, obj, selection: Selection, columns=None) -> DataFrame:
...
-                            # if we have a multi-index, then need to include
-                            # the levels
-                            if self.is_multi_index:
-                                filt = filt.union(Index(self.levels))
+                    # if we have a multi-index, then need to include
+                    # the levels
+                    if self.is_multi_index:
+                        filt = filt.union(Index(self.levels))
+                    takers = op(axis_values, filt)
+                    return obj.loc(axis=axis_number)[takers]
```
Comment (MarcoGorelli) [inline pandas/io/pytables.py:4127]: nit - let's not remove this newline

other than that, looks good

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/io/pytables.py (hunk 3 above)
Comment (roadswitcher) [reply pandas/io/pytables.py]: Whoops.      replaced newline, `git push`, let the tests begin.

File: pandas/io/pytables.py [hunk 4]
```diff
@@ -4106,43 +4106,43 @@ def process_axes(self, obj, selection: Selection, columns=None) -> DataFrame:
...
+                        takers = op(values, filt)
+                        return obj.loc(axis=axis_number)[takers]
 
-                    raise ValueError(f"cannot find the field [{field}] for filtering!")
+                raise ValueError(f"cannot find the field [{field}] for filtering!")
 
+        # apply the selection filters (but keep in the same order)
+        if selection.filter is not None:
+            for field, op, filt in selection.filter.format():
                 obj = process_filter(field, filt)
```
Comment (MarcoGorelli) [inline pandas/io/pytables.py:4146]: doesn't `process_filter` need to take `op` as an argument now? `obj = process_filter(field, filt, op)`

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/io/pytables.py (hunk 4 above)
Comment (roadswitcher) [reply pandas/io/pytables.py]: Yeah, when I replaced the line break, I replaced it with the original unchanged version of the file.

File: pandas/tests/io/parser/test_c_parser_only.py [hunk 5]
```diff
@@ -176,6 +176,9 @@ def test_precise_conversion(c_parser_only):
     normal_errors = []
     precise_errors = []
 
+    def error(val, actual_val):
+        return abs(Decimal(f"{val:.100}") - actual_val)
```
Comment (MarcoGorelli) [inline pandas/tests/io/parser/test_c_parser_only.py:180]: can we add type hints (both here and to the `_isin` function you've added) whilst we're here? any new development should ideally include them

other than that, looks good to me

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/tests/io/parser/test_c_parser_only.py (hunk 5 above)
Comment (roadswitcher) [reply pandas/tests/io/parser/test_c_parser_only.py]: Changed, ran `pre-commit`, pushed change.

File: pandas/core/strings/object_array.py [hunk 6]
```diff
@@ -380,9 +381,14 @@ def _str_get_dummies(self, sep: str = "|"):
 
         dummies = np.empty((len(arr), len(tags2)), dtype=np.int64)
 
+        def _isin(test_elements: Series, element: str) -> bool:
```
Comment (MarcoGorelli) [inline pandas/core/strings/object_array.py:384]:
```suggestion
        def _isin(test_elements: str, element: str) -> bool:
```
(try inserting a breakpoint or a print statement and see)

File: pandas/tests/io/parser/test_c_parser_only.py [hunk 7]
```diff
@@ -176,6 +176,9 @@ def test_precise_conversion(c_parser_only):
     normal_errors = []
     precise_errors = []
 
+    def error(val: float, actual_val: Decimal) -> float:
```
Comment (MarcoGorelli) [inline pandas/tests/io/parser/test_c_parser_only.py:179]:
```suggestion
    def error(val: float, actual_val: Decimal) -> Decimal:
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
almost there 💪

**Comment by roadswitcher** [comment]
> almost there 💪

The reward for a steep climb is the view from the top.

**Review by MarcoGorelli — APPROVED** [review]
Thanks @roadswitcher

**Pull request merged**